client/landing/stepper/declarative-flow/internals/steps-repository/sensei-launch/use-sub-steps.ts

```diff
@@ -0,0 +1,58 @@
+import { useEffect, useState } from 'react';
+
+export const wait = ( ms: number ) => new Promise( ( res ) => setTimeout( res, ms ) );
+
+/**
+ * A step in a linear installation process.
+ *
+ * @param {number} retries Number of times this step was executed.
+ *
+ * @returns Should resolve to true if the step was completed, false if it needs to be run again.
+ */
+type SubStep = ( retries: number ) => Promise< boolean >;
+
+type StepStatus = {
+	current: number;
+	retries: number;
+};
+
+/**
+ * Run a series of steps in order, re-running each step until it's complete.
+ *
+ * Steps should take care of waiting for requests or time delays, and resolve the returned promise
+ * to false when they are ready to run again, or true when they are complete and we can proceed
+ * with the next step.
+ *
+ * @param {SubStep[]} steps
+ * @returns {number} Progress percentage.
+ */
+export const useSubSteps = ( steps: SubStep[] ) => {
```

Comment:
Not sure if you intend to add it but I think this method should define a maximum number of retries and execute the last step if this is reached.

Reply:
Good idea, added back the max retries logic